output/cloud/expv2/output.go

```diff
@@ -51,14 +93,100 @@ func (o *Output) Start() error {
 // StopWithTestError gracefully stops all metric emission from the output.
 func (o *Output) StopWithTestError(testErr error) error {
 	o.logger.Debug("Stopping...")
+	close(o.stopMetricsCollection)
 
-	// TODO: add stop implementation
+	// Drain the SampleBuffer and force the aggregation for flushing
+	// all the queued samples even if they haven't yet passed the
+	// wait period.
+	o.periodicCollector.Stop()
+	o.collector.DropExpiringDelay()
+	o.collector.CollectSamples(nil)
+	o.periodicFlusher.Stop()
 
 	o.logger.Debug("Stopped!")
 	return nil
 }
 
 // AddMetricSamples receives the samples streaming.
 func (o *Output) AddMetricSamples(s []metrics.SampleContainer) {
-	// TODO: implement
+	// TODO: this and the next operation are two locking operations,
+	// evaluate to do something smarter, maybe having a lock-free
+	// queue.
+	select {
+	case <-o.stopMetricsCollection:
+		return
+	default:
+	}
+
+	// TODO: when we will have a very good optimized
+	// bucketing process we may evaluate to drop this
+	// buffer.
+	//
+	// If the bucketing process is efficient, the single
+	// operation could be a bit longer than just enqueuing
+	// but it could be fast enough to justify to to direct
+	// run it and save some memory across the e2e operation.
+	//
+	// It requires very specific benchmark.
+	o.SampleBuffer.AddMetricSamples(s)
+}
+
+func (o *Output) collectSamples() {
+	samples := o.GetBufferedSamples()
+	if len(samples) < 1 {
+		return
+	}
+	o.collector.CollectSamples(samples)
+
+	// TODO: other operations with the samples containers
+	// e.g. flushing the Metadata as tracing samples
+}
+
+// flushMetrics receives a set of metric samples.
+func (o *Output) flushMetrics() {
+	start := time.Now()
+
+	ctx, cancel := context.WithTimeout(context.Background(), o.config.MetricPushInterval.TimeDuration())
+	defer cancel()
+
+	err := o.metricsFlusher.Flush(ctx)
+	if err != nil {
```

Comment:
Minor non-blocking nit: you can probably flip this `if` around and have something like this:
```go
if err == nil {
	o.logger.WithField("t", time.Since(start)).Debug("Successfully flushed buffered samples to the cloud")
	return
}
// ...long nested logic...
```

Reply:
This is one line after the refactor in this commit in the next PR https://github.com/grafana/k6/pull/3082/commits/8a9acd673f7191474db008924dc4d958d40a0311